client-go/util/jsonpath: resolve #16707 by outputting json for non-primitive types

pjferrell opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When printing JSONPATH results, this changes the display format of non-primitive types from Go-syntax to JSON.  This improves readability for maps/slices and simplifies processing outputs in other tools.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #16707
Fixes #67268

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
jsonpath support in kubectl / client-go serializes complex types (maps / slices / structs) as json instead of Go-syntax.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pjferrell commented:
/assign @smarterclayton

brianpursley commented:
/ok-to-test

brianpursley commented:
Hi @pjferrell 
Thanks for the PR.

if you edit your original message to change `Fixes # 16707` to `Fixes #16707` (remove the space) it will turn that into a link to the issue which will help the reviewers.

brianpursley commented on staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:
```diff
@@ -121,7 +122,14 @@ func (j *JSONPath) FindResults(data interface{}) ([][]reflect.Value, error) {
+		var text []byte
+		var err error
+		if r.Kind() == reflect.Interface && (r.Elem().Kind() == reflect.Map ||
+			r.Elem().Kind() == reflect.Slice || r.Elem().Kind() == reflect.Struct) {
```
Do you want this change to apply for slice and struct?  I thought the issue was only about map.  I guess I would need to see some unit tests around those types to understand how the output would be different.

pjferrell replied:
Based on the comments in #16707 (@thockin last comment and others) indicated it'd be favorable to address all non-primitive types which could be represented by json.  Go-style output of structs does not contain attribute/key names and slices are printed without comma separators, which makes passing these outputs to other tools more difficult.

pjferrell replied:
The current test failure is an example jsonpath output where the slice was previously output is Go-style instead of json.

Expected was the old output, and actual is the proper json representation of the slice.
```
--- FAIL: TestTokenOutput (0.00s)
    --- PASS: TestTokenOutput/JSON_output (0.00s)
    --- PASS: TestTokenOutput/YAML_output (0.00s)
    --- PASS: TestTokenOutput/Go_template_output (0.00s)
    --- PASS: TestTokenOutput/text_output (0.00s)
    --- FAIL: TestTokenOutput/jsonpath_output (0.00s)
        token_test.go:449: failed TestTokenOutput:
            
            expected:
            abcdef.1234567890123456 [system:bootstrappers:kubeadm:default-node-token]
            
            actual:
            abcdef.1234567890123456 ["system:bootstrappers:kubeadm:default-node-token"]
```

brianpursley replied:
Can you add a couple unit tests to cover the slice and struct scenarios.
Something like this might work to cover both (slice with multiple structs):

```
{range .items[1]}{.status.addresses} {end}
```

I'm just thinking it will be good to have this to show the expected output and prevent any future regression.

pjferrell replied:
I've added several test cases for each primitive type.  While adding test cases, I resolved some missing/broken filter behavior that was skipping certain types / unstable sorting.

brianpursley commented:
/retest

pjferrell commented:
/retest

brianpursley commented:
Suggestion on your release note, you might want to word it in a way that assumes it will be in a long list of changes.  When you say "this fix" it won't mean anything to someone reading the Kubernetes release notes because it is out of context.  

> When using the JSONPATH output option in kubectl (or client-go/util/jsonpath package), complex types (maps / slices / structs) were previously displayed as Go-syntax.  This fix changes the behavior to return json for non-primitive types.

I suggest rewording it to a single sentence if possible and just describe the new kubectl behavior without making any reference to this PR or the Issue.

pjferrell commented:
/retest

pjferrell commented:
/retest

brianpursley commented:
Wow this is getting to be a pretty big PR compared to the original commit.  Maybe consider breaking the work up into multiple PRs will help the reviewers be able to approve more easily.

pjferrell committed:
client-go: change jsonpath output of non-primitive types from Go-syntax to JSON
kubectl: add --output jsonpath-as-json to print a json array of results
kubeadm: fix test case using jsonpath result of non-primitive type

pjferrell commented:
@brianpursley I pulled out the code for providing a stable/sorted jsonpath recursion/iteration (primarily affects recursion of maps).

The remaining code is the minimal changes to achieve:
1. client-go/util/jsonpath: output non-primitives (slice/struct/map) as json
2. kubectl: add --output=jsonpath-as-json to print a json list of results (ex. slice of Ingresses, slice of Containers, slice of .metadata.name, etc)
3. kubeadm: fix test case that was expecting Go-style output of slice

I'll create a new PR for stable/sorted jsonpath recursion/iteration once this PR is merged.  It will only affect client-go/util/jsonpath package

pjferrell commented:
/test pull-kubernetes-integration

pjferrell commented:
/test pull-kubernetes-e2e-gce

pjferrell commented:
Related PRs that were closed:
https://github.com/kubernetes/kubernetes/pull/26099
https://github.com/kubernetes/kubernetes/pull/33890

pjferrell commented:
/assign timothysc
/assign seans3

neolit123 commented:
/approve
for cmd/kubeadm

/area kubectl
/priority backlog
(given this has been an issue for a while)

pjferrell committed:
apiextensions-apiserver: update test for CRD custom columns to jsonpath output of non-primitive type

lavalamp commented:
/assign

lavalamp commented:
As long as this is a client-side change, which it seems to be, I'm fine with it (thanks for all the tests). I think someone from SIG CLI should sign off on it, then I'm happy to approve for the api machinery OWNERs files.

/assign @pwittrock @seans3 @apelisse

pjferrell commented:
@lavalamp It does change the default output behavior for complex objects to json instead of go-syntax.  IMO it would be a positive change, but it could break a script expecting go-style output ("1 2 3" vs "[1,2,3]").  I fixed all kubernetes test cases (3) that used the go-style output.

The cli flag provides an output option to display a json slice of results (consistent with how other jsonpath implementations output).

lavalamp commented:
Yeah, I think that makes sense, we just need SIG CLI to agree.

On Tue, Apr 28, 2020 at 1:54 PM Phillip Ferrell <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> It does change the default output
> behavior for complex objects to json instead of go-syntax. IMO it would be
> a positive change, but it could break a script expecting go-style output
> ("1 2 3" vs "[1,2,3]"). I fixed all kubernetes test cases (3) that used the
> go-style output.
>
> The cli flag provides an output option to display a json slice of results
> (consistent with how other jsonpath implementations output).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/89660#issuecomment-620849272>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFX4S3KB66UIPT7EXA3RO47B5ANCNFSM4LWZNMLQ>
> .
>

apelisse commented:
I see a couple of comments from informed people that make sense to me:

From https://github.com/kubernetes/kubernetes/pull/33890#issuecomment-251278083:
> Single values are used to provide lists of arguments. Changing the format would break all scripts using it.
>
> The printing of complex values is a bug, and can be fixed.

From https://github.com/kubernetes/kubernetes/pull/33890#issuecomment-252347667:
> The go type dumps are borderline useless. We can either say JSON or we can
define our own format.

lavalamp commented:
/approve
/lgtm

OK that's good enough for me

pjferrell commented:
/retest

thockin commented:
This is awesome!

Now how do I format a JSON list without a trailing comma?

```
kubectl get nodes -o jsonpath='{"{"}{range .items[*]}"{.metadata.name}": {.metadata.labels},{end}{"}"}' | jq .
parse error: Expected another key-value pair at line 1, column 1582
```

:P

brianpursley commented:
Its probably not what you're looking for, but you *could* use sed to remove it...
```
kubectl get nodes -o jsonpath='{"{"}{range .items[*]}"{.metadata.name}": {.metadata.labels},{end}{"}"}' | sed 's/\(.*\),/\1/' | jq
```

Are you thinking there should be some kind of conditional rendering? ...maybe some new thing like `{separator ","}` that conditionally renders only when it is a separating range iterations? 

Is jsonpath an actual spec?

Also, if it were able to render \b, then you could add a `\b ` after the last `{end}` to "backspace away" the last comma.  A little hacky but might be a simple way to do it if it actually worked.

lavalamp commented:
IMO just use jq if you want to do that, it's much simpler.

To do it without jq, I would do something exceedingly stupid, e.g.

```sh
$ echo "{x,y,}" | (head -c -3; echo "}")
{x,y}
```

lavalamp commented:
I haven't tested it but a jq script should look something like this:

`jq '[.items[*] | {.metadata.name, .metadata.labels}]'`

(Yes, I feel guilty for being this nerdy about jq)
